src/utils/unit_test/fftw_test.cpp

```diff
@@ -0,0 +1,401 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+// MUST include this
+#include <catch2/catch.hpp>
+
+#include <lbann/base.hpp>
+#include <lbann/utils/exception.hpp>
+#include <lbann/utils/fftw_wrapper.hpp>
+
+#include <complex>
+#include <iostream>
+#include <vector>
+
+#include <fftw3.h>
+
+// These
+namespace
+{
+auto get_input_dims(size_t ndims)
+{
+  std::vector<int> dims(ndims+1, 16);
+  dims.front() = 4;// 4 channels of size 16^ndims
+  return dims;
+}
+
+auto get_num_samples()
+{
+  return 7;
+}
+
+auto get_ldim_offset()
+{
+  return 13;// because PRIME!
+}
+
+template <typename T>
+bool assert_r2c_symmetry_1d(
+  El::Matrix<T, El::Device::CPU> const& full_output,
+  std::vector<int> const& full_dims)
+{
+  if (full_dims.size() != 2UL)
+    LBANN_ERROR("Only valid for 1-D feature maps.");
+  auto const num_feat_maps = full_dims[0];
+  auto const num_entries = full_dims[1];
+  auto const num_samples = full_output.Width();
+
+  auto const r2c_dims = lbann::fft::get_r2c_output_dims(full_dims);
+  auto const num_entries_r2c = r2c_dims[1];
+
+  bool all_good = true;
+  for (El::Int sample = 0; sample < num_samples; ++sample)
+  {
+    for (int feat_map = 0; feat_map < num_feat_maps; ++feat_map)
+    {
+      T const* feat_map_mat = full_output.LockedBuffer()
+        + sample*full_output.LDim()
+        + feat_map*num_entries;
+      for (int ent = num_entries_r2c; ent < num_entries; ++ent)
+      {
+        auto const conj_ent = (ent == 0 ? 0 : num_entries-ent);
+        auto const val = feat_map_mat[ent];
+        auto const conj_val = feat_map_mat[conj_ent];
+        if (val != conj_val)
```

Comment:
Shouldn't we be checking they are conjugates?

Reply:
This was some debugging code that never got cleaned up. I'll do a quick tidy before merge.

Reply:
Done